pretty: use combining mark aware centering and wrapping

pyBlob opened the pull request:
The console pprint-function does not insert the correct amount of spaces in an expression like:
```
>>> pprint([Symbol("ihat") / (Symbol("i") + 1)])
⎡  î ⎤
⎢─────⎥
⎣i + 1⎦
```

While it does for following expression:
```
>>> pprint(Matrix([Symbol("ihat"), Symbol("i") + 1]))
⎡  î  ⎤
⎢     ⎥
⎣i + 1⎦
```

The difference is that the centering done by `PrettyPrinter._print_matrix_contents` is unicode-aware, while the centering of `stringPict.equalLengths` and `stringPict.stack` is not.

The method `stringPict.render` also uses indexing to wrap lines. This messes up the alignments between individual lines of wrapped sub-pictures. This may even remove some of the trailing marks when the wrapping separates the character and its mark.

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

- Fixes #9450

#### Brief description of what is fixed or changed

- Extract centering code of `PrettyPrinter._print_matrix_contents` to `pretty_symbology.center_pad`.
- Reuse for `stringPict.equalLengths` and `stringPict.stack` .
- Rewrite wrapping logic of `stringPict.render`:
  - Handle each character individually
  - Insert overflow indicators like "vim" and "less" do for long lines.

#### Other comments

TODO: Should I add the example at the top to the docstring of `center` or `center_pad`? Does it automatically run on test?

TODO: Fix broken whitespace in testcases + examples.

Why do the failing examples lack trailing whitespace? This PR should not have had any impact on those examples, because they don't use combining marks.

Example: (Select following code block and note the missing whitespace in "Expected")
```
sympy/sympy/solvers/ode/single.py", line 1634, in sympy.solvers.ode.single.HomogeneousCoeffSubsIndepDivDep
Failed example:
    pprint(dsolve(2*x*f(x) + (x**2 + f(x)**2)*f(x).diff(x), f(x),
    hint='1st_homogeneous_coeff_subs_indep_div_dep',
    simplify=False))
Expected:
                             /    2    \
                             | 3*x     |
                          log|----- + 1|
                             | 2       |
                             \f (x)    /
    log(f(x)) = log(C1) - --------------
                                3
Got:
                             /   2     \
                             |3*x      |
                          log|----- + 1|
                             | 2       |
                             \f (x)    /
    log(f(x)) = log(C1) - --------------
                                3       
```

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* printing
  * pretty: use combining mark aware centering for equalLengths and stack
  * pretty: use combining mark aware wrapping
<!-- END RELEASE NOTES -->

Commit:
pretty: use combining mark aware centering for equalLengths and stack

Commit:
pretty: use combining mark aware wrapping

oscargus commented:
Thank you for this contribution! Looks promising!

> Should I add the example at the top to the docstring of `center` or `center_pad`? 

I do not think it is needed.

> Does it automatically run on test?

Yes, it is executed as part of the doc-tests, but the main purpose of the doc-test is to make sure that the docs are correct, not that the function is correct. So please add tests in `printing/pretty/tests/test_pretty.py`.

smichr committed:
Merge branch 'master' into fix/combing_marks_centering

smichr commented on sympy/printing/pretty/stringpict.py:
```diff
@@ -262,38 +259,63 @@ def render(self, * args, **kwargs):
+        """
+        Break long-lines in a visually pleasing format.
+        without overflow indicators | with overflow indicators
+        
```
```suggestion

```

smichr committed:
Update sympy/printing/pretty/stringpict.py

smichr committed:
Update test_pretty.py

smichr commented on sympy/printing/pretty/stringpict.py:
```diff
@@ -262,38 +259,62 @@ def render(self, * args, **kwargs):
+        |                     |     |                     |
+        |     3    4    4     |     |↪      3    4    4   |
+        |4*y*x  + x  + y      |     |↪ 4*y*x  + x  + y    |
+        
```
```suggestion

```

smichr committed:
Update sympy/printing/pretty/stringpict.py

smichr committed:
whitelist stringpict

smichr commented on sympy/printing/pretty/tests/test_pretty.py:
```diff
@@ -3021,6 +3022,14 @@ def test_pretty_derivatives():
+-----\\x  + log(x)/\n\
+dy dx             \
+"""
+    
```
```suggestion
```

smichr committed:
Update sympy/printing/pretty/tests/test_pretty.py

smichr committed:
Update test_pretty.py

smichr committed:
try remove gremlin

Ran black and copied 5000 lines and pasted and then copied and pasted the remaining lines. Not sure why  the invalid escape sequence was showing at 5620 before this commit.

smichr committed:
restore and fix escape error

smichr committed:
Update test_pretty.py

smichr commented:
I will put up a replacement PR today that continues this.
